File: backend/app/api/routes/employee_customers.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.core.permissions import require_employee
from app.database.dependencies import get_db
from app.models.user import User
from app.schemas.employee_customer import (
    EmployeeCustomerCreate,
    EmployeeCustomerCreateResponse,
    EmployeeCustomerResponse,
    EmployeeCustomerResetPasswordResponse,
    EmployeeCustomerStatusUpdate,
    EmployeeCustomerUpdate,
)
from app.services.employee_customer_service import (
    EmployeeCustomerService,
)
from app.repositories.employee_customer_repository import (
    EmployeeCustomerRepository,
)
from app.schemas.citizen_profile import (
    CitizenProfileCreate,
    CitizenProfileResponse,
    CitizenProfileUpdate,
)
from app.services.citizen_profile_service import CitizenProfileService
from app.services.rule_engine_service import RuleEngineService
from app.services.ai_recommendation_service import AIRecommendationService
from app.schemas.application import ApplicationResponse
from app.services.application_service import ApplicationService

router = APIRouter(
    prefix="/employee/customers",
    tags=["Employee Customers"],
)


@router.get(
    "/",
    response_model=list[EmployeeCustomerResponse],
)
def get_customers(
    search: str | None = Query(
        default=None,
        description="Search by customer name or email.",
    ),
    status_filter: str = Query(
        default="all",
        alias="status",
        pattern="^(all|active|inactive)$",
    ),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    return EmployeeCustomerService.get_customers(
        db=db,
        search=search,
        status_filter=status_filter,
    )


@router.post(
    "/",
    response_model=EmployeeCustomerCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_customer(
    payload: EmployeeCustomerCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    try:
        customer, temporary_password = (
            EmployeeCustomerService.create_customer(
                db=db,
                full_name=payload.full_name,
                email=payload.email,
            )
        )

        return {
            "customer": customer,
            "temporary_password": temporary_password,
        }

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

@router.get(
    "/{customer_id}/profile",
    response_model=CitizenProfileResponse,
)
def get_customer_profile(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerRepository.get_customer_by_id(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    profile = CitizenProfileService.get_profile(
        db=db,
        user_id=customer.id,
    )

    if not profile:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Citizen profile not found.",
        )

    return profile

@router.post(
    "/{customer_id}/profile",
    response_model=CitizenProfileResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_customer_profile(
    customer_id: int,
    profile: CitizenProfileCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerRepository.get_customer_by_id(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    try:
        return CitizenProfileService.create_profile(
            db=db,
            user_id=customer.id,
            profile=profile,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

@router.put(
    "/{customer_id}/profile",
    response_model=CitizenProfileResponse,
)
def update_customer_profile(
    customer_id: int,
    profile: CitizenProfileUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerRepository.get_customer_by_id(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    try:
        return CitizenProfileService.update_profile(
            db=db,
            user_id=customer.id,
            profile=profile,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )

@router.delete(
    "/{customer_id}/profile",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_customer_profile(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerRepository.get_customer_by_id(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    try:
        CitizenProfileService.delete_profile(
            db=db,
            user_id=customer.id,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )

@router.get(
    "/{customer_id}/eligibility",
)
def get_customer_eligibility(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerRepository.get_customer_by_id(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    try:
        schemes = RuleEngineService.get_eligible_schemes(
            db=db,
            user_id=customer.id,
        )

        return {
            "eligible_count": len(schemes),
            "eligible_schemes": [
                {
                    "id": scheme.id,
                    "scheme_name": scheme.scheme_name,
                    "category": scheme.category,
                    "department": scheme.department,
                    "description": scheme.description,
                    "official_url": scheme.official_url,
                }
                for scheme in schemes
            ],
        }

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )


@router.get(
    "/{customer_id}/recommendations",
)
def get_customer_recommendations(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerRepository.get_customer_by_id(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    try:
        recommendations = AIRecommendationService.get_recommendations(
            db=db,
            user_id=customer.id,
        )

        return {
            "recommendation_count": len(recommendations),
            "recommendations": recommendations,
        }

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )


@router.get(
    "/{customer_id}/applications",
    response_model=list[ApplicationResponse],
)
def get_customer_applications(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    try:
        return ApplicationService.get_customer_applications(
            db=db,
            customer_id=customer_id,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )


@router.get(
    "/{customer_id}",
    response_model=EmployeeCustomerResponse,
)
def get_customer(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    customer = EmployeeCustomerService.get_customer(
        db=db,
        customer_id=customer_id,
    )

    if not customer:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Customer not found.",
        )

    return customer


@router.put(
    "/{customer_id}",
    response_model=EmployeeCustomerResponse,
)
def update_customer(
    customer_id: int,
    payload: EmployeeCustomerUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    try:
        return EmployeeCustomerService.update_customer(
            db=db,
            customer_id=customer_id,
            full_name=payload.full_name,
            email=payload.email,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )


@router.patch(
    "/{customer_id}/status",
    response_model=EmployeeCustomerResponse,
)
def update_customer_status(
    customer_id: int,
    payload: EmployeeCustomerStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    try:
        return EmployeeCustomerService.update_customer_status(
            db=db,
            customer_id=customer_id,
            is_active=payload.is_active,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )


@router.post(
    "/{customer_id}/reset-password",
    response_model=EmployeeCustomerResetPasswordResponse,
)
def reset_customer_password(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    try:
        customer, temporary_password = (
            EmployeeCustomerService.reset_customer_password(
                db=db,
                customer_id=customer_id,
            )
        )

        return {
            "customer": customer,
            "temporary_password": temporary_password,
        }

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )


@router.delete(
    "/{customer_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_customer(
    customer_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_employee),
):
    try:
        EmployeeCustomerService.delete_customer(
            db=db,
            customer_id=customer_id,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )

File: frontend/src/api/employeeCustomers.ts
import { apiClient } from "./client";

import type {
  CitizenProfile,
  CitizenProfilePayload,
  EligibilityResponse,
  Application,
  RecommendationsResponse,
} from "../types/api";

export interface EmployeeCustomer {
  id: number;
  full_name: string;
  email: string;
  role: string;
  is_active: boolean;
  must_change_password: boolean;
}

export interface CreateCustomerPayload {
  full_name: string;
  email: string;
}

export interface CreateCustomerResponse {
  customer: EmployeeCustomer;
  temporary_password: string;
}

export interface UpdateCustomerPayload {
  full_name: string;
  email: string;
}

export interface CustomerStatusPayload {
  is_active: boolean;
}

export interface ResetCustomerPasswordResponse {
  customer: EmployeeCustomer;
  temporary_password: string;
}

export async function getEmployeeCustomers(
  search = "",
  status: "all" | "active" | "inactive" = "all",
): Promise<EmployeeCustomer[]> {
  const { data } = await apiClient.get<EmployeeCustomer[]>(
    "/employee/customers/",
    {
      params: {
        search: search.trim() || undefined,
        status,
      },
    },
  );

  return data;
}

export async function getEmployeeCustomer(
  customerId: number,
): Promise<EmployeeCustomer> {
  const { data } = await apiClient.get<EmployeeCustomer>(
    `/employee/customers/${customerId}`,
  );

  return data;
}

export async function createEmployeeCustomer(
  payload: CreateCustomerPayload,
): Promise<CreateCustomerResponse> {
  const { data } =
    await apiClient.post<CreateCustomerResponse>(
      "/employee/customers/",
      payload,
    );

  return data;
}

export async function updateEmployeeCustomer(
  customerId: number,
  payload: UpdateCustomerPayload,
): Promise<EmployeeCustomer> {
  const { data } =
    await apiClient.put<EmployeeCustomer>(
      `/employee/customers/${customerId}`,
      payload,
    );

  return data;
}

export async function updateEmployeeCustomerStatus(
  customerId: number,
  payload: CustomerStatusPayload,
): Promise<EmployeeCustomer> {
  const { data } =
    await apiClient.patch<EmployeeCustomer>(
      `/employee/customers/${customerId}/status`,
      payload,
    );

  return data;
}

export async function resetEmployeeCustomerPassword(
  customerId: number,
): Promise<ResetCustomerPasswordResponse> {
  const { data } =
    await apiClient.post<ResetCustomerPasswordResponse>(
      `/employee/customers/${customerId}/reset-password`,
    );

  return data;
}

export async function deleteEmployeeCustomer(
  customerId: number,
): Promise<void> {
  await apiClient.delete(
    `/employee/customers/${customerId}`,
  );
}

export async function getEmployeeCustomerProfile(
  customerId: number,
): Promise<CitizenProfile> {
  const { data } = await apiClient.get<CitizenProfile>(
    `/employee/customers/${customerId}/profile`,
  );

  return data;
}

export async function createEmployeeCustomerProfile(
  customerId: number,
  payload: CitizenProfilePayload,
): Promise<CitizenProfile> {
  const { data } =
    await apiClient.post<CitizenProfile>(
      `/employee/customers/${customerId}/profile`,
      payload,
    );

  return data;
}

export async function updateEmployeeCustomerProfile(
  customerId: number,
  payload: Partial<CitizenProfilePayload>,
): Promise<CitizenProfile> {
  const { data } =
    await apiClient.put<CitizenProfile>(
      `/employee/customers/${customerId}/profile`,
      payload,
    );

  return data;
}

export async function getEmployeeCustomerEligibility(
  customerId: number,
): Promise<EligibilityResponse> {
  const { data } =
    await apiClient.get<EligibilityResponse>(
      `/employee/customers/${customerId}/eligibility`,
    );

  return data;
}


export async function getEmployeeCustomerRecommendations(
  customerId: number,
): Promise<RecommendationsResponse> {
  const { data } =
    await apiClient.get<RecommendationsResponse>(
      `/employee/customers/${customerId}/recommendations`,
    );

  return data;
}

export async function getEmployeeCustomerApplications(
  customerId: number,
): Promise<Application[]> {
  const { data } =
    await apiClient.get<Application[]>(
      `/employee/customers/${customerId}/applications`,
    );

  return data;
}
